bin/camelot.ml

```diff
@@ -51,34 +54,44 @@ let lex_src file =
 let parse_src (src, lexbuf) =
   src, Parse.implementation lexbuf
 
+let sanitize_dir d =
+    if d.[String.length d - 1] = '/' then d 
+    else d ^ "/"
+
 let files_in_dir dirname = 
   let open Sys in
-  let sanitize_dir d =
-    if d.[String.length d - 1] = '/' then d 
-    else d ^ "/" in
   let dir = sanitize_dir dirname in
   if not (file_exists dir && is_directory dir) 
   then fail @@ dir ^ " doesn't exist or isn't a directory!";
   readdir dir |> Array.to_list |> List.map (fun file -> dir ^ file)
 
 let usage_msg =
+  "invoke with -r (only works if -d is set too) to recurse into subdirectories\n" ^
   "invoke with -d <dir_name> to specify a directory to lint, or just run the program with default args\n" ^
   "invoke with -show <student | ta | gradescope> to select the display type - usually ta's want a briefer summary" ^
   "invoke with -f <.ml filename> to lint a particular file"
 
-let parse_sources_in dirname = 
+let rec parse_sources_in dirname : (string * Parsetree.structure) list = 
```

Comment:
Separate the functionality of recursively finding for file names from actually lexing and parsing the files.

Perhaps make a ```let files_in_dir_rec ``` helper / merge with the files_in_dir function? Then use that inside of parse_sources_in, to determine the files that have to be parsed. This just helps separate the fxnality of parsing from the functionality of finding files :D

Reply:
okay that sounds appropriate i'll implement this

Reply:
done